File: OpenHome/Net/ControlPoint/Proxies/CpAvOpenhomeOrgNetworkMonitor1.cpp
#include "CpAvOpenhomeOrgNetworkMonitor1.h"
#include <OpenHome/Net/Core/CpProxy.h>
#include <OpenHome/Net/Private/CpiService.h>
#include <OpenHome/Private/Thread.h>
#include <OpenHome/Net/Private/AsyncPrivate.h>
#include <OpenHome/Net/Core/CpDevice.h>
#include <OpenHome/Net/Private/Error.h>
#include <OpenHome/Net/Private/CpiDevice.h>

using namespace OpenHome;
using namespace OpenHome::Net;


class SyncNameAvOpenhomeOrgNetworkMonitor1 : public SyncProxyAction
{
public:
    SyncNameAvOpenhomeOrgNetworkMonitor1(CpProxyAvOpenhomeOrgNetworkMonitor1& aProxy, Brh& aName);
    virtual void CompleteRequest(IAsync& aAsync);
    virtual ~SyncNameAvOpenhomeOrgNetworkMonitor1() {}
private:
    CpProxyAvOpenhomeOrgNetworkMonitor1& iService;
    Brh& iName;
};

SyncNameAvOpenhomeOrgNetworkMonitor1::SyncNameAvOpenhomeOrgNetworkMonitor1(CpProxyAvOpenhomeOrgNetworkMonitor1& aProxy, Brh& aName)
    : iService(aProxy)
    , iName(aName)
{
}

void SyncNameAvOpenhomeOrgNetworkMonitor1::CompleteRequest(IAsync& aAsync)
{
    iService.EndName(aAsync, iName);
}


class SyncPortsAvOpenhomeOrgNetworkMonitor1 : public SyncProxyAction
{
public:
    SyncPortsAvOpenhomeOrgNetworkMonitor1(CpProxyAvOpenhomeOrgNetworkMonitor1& aProxy, TUint& aSender, TUint& aReceiver, TUint& aResults);
    virtual void CompleteRequest(IAsync& aAsync);
    virtual ~SyncPortsAvOpenhomeOrgNetworkMonitor1() {}
private:
    CpProxyAvOpenhomeOrgNetworkMonitor1& iService;
    TUint& iSender;
    TUint& iReceiver;
    TUint& iResults;
};

SyncPortsAvOpenhomeOrgNetworkMonitor1::SyncPortsAvOpenhomeOrgNetworkMonitor1(CpProxyAvOpenhomeOrgNetworkMonitor1& aProxy, TUint& aSender, TUint& aReceiver, TUint& aResults)
    : iService(aProxy)
    , iSender(aSender)
    , iReceiver(aReceiver)
    , iResults(aResults)
{
}

void SyncPortsAvOpenhomeOrgNetworkMonitor1::CompleteRequest(IAsync& aAsync)
{
    iService.EndPorts(aAsync, iSender, iReceiver, iResults);
}


CpProxyAvOpenhomeOrgNetworkMonitor1::CpProxyAvOpenhomeOrgNetworkMonitor1(CpDevice& aDevice)
    : CpProxy("av-openhome-org", "NetworkMonitor", 1, aDevice.Device())
{
    OpenHome::Net::Parameter* param;

    iActionName = new Action("Name");
    param = new OpenHome::Net::ParameterString("Name");
    iActionName->AddOutputParameter(param);

    iActionPorts = new Action("Ports");
    param = new OpenHome::Net::ParameterUint("Sender");
    iActionPorts->AddOutputParameter(param);
    param = new OpenHome::Net::ParameterUint("Receiver");
    iActionPorts->AddOutputParameter(param);
    param = new OpenHome::Net::ParameterUint("Results");
    iActionPorts->AddOutputParameter(param);

    Functor functor;
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgNetworkMonitor1::NamePropertyChanged);
    iName = new PropertyString(aDevice.Device().GetCpStack().Env(), "Name", functor);
    AddProperty(iName);
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgNetworkMonitor1::SenderPropertyChanged);
    iSender = new PropertyUint(aDevice.Device().GetCpStack().Env(), "Sender", functor);
    AddProperty(iSender);
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgNetworkMonitor1::ReceiverPropertyChanged);
    iReceiver = new PropertyUint(aDevice.Device().GetCpStack().Env(), "Receiver", functor);
    AddProperty(iReceiver);
    functor = MakeFunctor(*this, &CpProxyAvOpenhomeOrgNetworkMonitor1::ResultsPropertyChanged);
    iResults = new PropertyUint(aDevice.Device().GetCpStack().Env(), "Results", functor);
    AddProperty(iResults);
}

CpProxyAvOpenhomeOrgNetworkMonitor1::~CpProxyAvOpenhomeOrgNetworkMonitor1()
{
    DestroyService();
    delete iActionName;
    delete iActionPorts;
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::SyncName(Brh& aName)
{
    SyncNameAvOpenhomeOrgNetworkMonitor1 sync(*this, aName);
    BeginName(sync.Functor());
    sync.Wait();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::BeginName(FunctorAsync& aFunctor)
{
    Invocation* invocation = iService->Invocation(*iActionName, aFunctor);
    TUint outIndex = 0;
    const Action::VectorParameters& outParams = iActionName->OutputParameters();
    invocation->AddOutput(new ArgumentString(*outParams[outIndex++]));
    iInvocable.InvokeAction(*invocation);
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::EndName(IAsync& aAsync, Brh& aName)
{
    ASSERT(((Async&)aAsync).Type() == Async::eInvocation);
    Invocation& invocation = (Invocation&)aAsync;
    ASSERT(invocation.Action().Name() == Brn("Name"));

    Error::ELevel level;
    TUint code;
    const TChar* ignore;
    if (invocation.Error(level, code, ignore)) {
        THROW_PROXYERROR(level, code);
    }
    TUint index = 0;
    ((ArgumentString*)invocation.OutputArguments()[index++])->TransferTo(aName);
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::SyncPorts(TUint& aSender, TUint& aReceiver, TUint& aResults)
{
    SyncPortsAvOpenhomeOrgNetworkMonitor1 sync(*this, aSender, aReceiver, aResults);
    BeginPorts(sync.Functor());
    sync.Wait();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::BeginPorts(FunctorAsync& aFunctor)
{
    Invocation* invocation = iService->Invocation(*iActionPorts, aFunctor);
    TUint outIndex = 0;
    const Action::VectorParameters& outParams = iActionPorts->OutputParameters();
    invocation->AddOutput(new ArgumentUint(*outParams[outIndex++]));
    invocation->AddOutput(new ArgumentUint(*outParams[outIndex++]));
    invocation->AddOutput(new ArgumentUint(*outParams[outIndex++]));
    iInvocable.InvokeAction(*invocation);
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::EndPorts(IAsync& aAsync, TUint& aSender, TUint& aReceiver, TUint& aResults)
{
    ASSERT(((Async&)aAsync).Type() == Async::eInvocation);
    Invocation& invocation = (Invocation&)aAsync;
    ASSERT(invocation.Action().Name() == Brn("Ports"));

    Error::ELevel level;
    TUint code;
    const TChar* ignore;
    if (invocation.Error(level, code, ignore)) {
        THROW_PROXYERROR(level, code);
    }
    TUint index = 0;
    aSender = ((ArgumentUint*)invocation.OutputArguments()[index++])->Value();
    aReceiver = ((ArgumentUint*)invocation.OutputArguments()[index++])->Value();
    aResults = ((ArgumentUint*)invocation.OutputArguments()[index++])->Value();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::SetPropertyNameChanged(Functor& aFunctor)
{
    iLock->Wait();
    iNameChanged = aFunctor;
    iLock->Signal();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::SetPropertySenderChanged(Functor& aFunctor)
{
    iLock->Wait();
    iSenderChanged = aFunctor;
    iLock->Signal();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::SetPropertyReceiverChanged(Functor& aFunctor)
{
    iLock->Wait();
    iReceiverChanged = aFunctor;
    iLock->Signal();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::SetPropertyResultsChanged(Functor& aFunctor)
{
    iLock->Wait();
    iResultsChanged = aFunctor;
    iLock->Signal();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::PropertyName(Brhz& aName) const
{
    AutoMutex a(PropertyReadLock());
    ASSERT(iCpSubscriptionStatus == CpProxy::eSubscribed);
    aName.Set(iName->Value());
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::PropertySender(TUint& aSender) const
{
    AutoMutex a(PropertyReadLock());
    ASSERT(iCpSubscriptionStatus == CpProxy::eSubscribed);
    aSender = iSender->Value();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::PropertyReceiver(TUint& aReceiver) const
{
    AutoMutex a(PropertyReadLock());
    ASSERT(iCpSubscriptionStatus == CpProxy::eSubscribed);
    aReceiver = iReceiver->Value();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::PropertyResults(TUint& aResults) const
{
    AutoMutex a(PropertyReadLock());
    ASSERT(iCpSubscriptionStatus == CpProxy::eSubscribed);
    aResults = iResults->Value();
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::NamePropertyChanged()
{
    ReportEvent(iNameChanged);
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::SenderPropertyChanged()
{
    ReportEvent(iSenderChanged);
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::ReceiverPropertyChanged()
{
    ReportEvent(iReceiverChanged);
}

void CpProxyAvOpenhomeOrgNetworkMonitor1::ResultsPropertyChanged()
{
    ReportEvent(iResultsChanged);
}

File: OpenHome/Net/Bindings/Cs/ControlPoint/Proxies/CpAvOpenhomeOrgNetworkMonitor1.cs
using System;
using System.Runtime.InteropServices;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using OpenHome.Net.Core;
using OpenHome.Net.ControlPoint;

namespace OpenHome.Net.ControlPoint.Proxies
{
    public interface ICpProxyAvOpenhomeOrgNetworkMonitor1 : ICpProxy, IDisposable
    {
        void SyncName(out String aName);
        void BeginName(CpProxy.CallbackAsyncComplete aCallback);
        void EndName(IntPtr aAsyncHandle, out String aName);
        void SyncPorts(out uint aSender, out uint aReceiver, out uint aResults);
        void BeginPorts(CpProxy.CallbackAsyncComplete aCallback);
        void EndPorts(IntPtr aAsyncHandle, out uint aSender, out uint aReceiver, out uint aResults);
        void SetPropertyNameChanged(System.Action aNameChanged);
        String PropertyName();
        void SetPropertySenderChanged(System.Action aSenderChanged);
        uint PropertySender();
        void SetPropertyReceiverChanged(System.Action aReceiverChanged);
        uint PropertyReceiver();
        void SetPropertyResultsChanged(System.Action aResultsChanged);
        uint PropertyResults();
    }

    internal class SyncNameAvOpenhomeOrgNetworkMonitor1 : SyncProxyAction
    {
        private CpProxyAvOpenhomeOrgNetworkMonitor1 iService;
        private String iName;

        public SyncNameAvOpenhomeOrgNetworkMonitor1(CpProxyAvOpenhomeOrgNetworkMonitor1 aProxy)
        {
            iService = aProxy;
        }
        public String Name()
        {
            return iName;
        }
        protected override void CompleteRequest(IntPtr aAsyncHandle)
        {
            iService.EndName(aAsyncHandle, out iName);
        }
    };

    internal class SyncPortsAvOpenhomeOrgNetworkMonitor1 : SyncProxyAction
    {
        private CpProxyAvOpenhomeOrgNetworkMonitor1 iService;
        private uint iSender;
        private uint iReceiver;
        private uint iResults;

        public SyncPortsAvOpenhomeOrgNetworkMonitor1(CpProxyAvOpenhomeOrgNetworkMonitor1 aProxy)
        {
            iService = aProxy;
        }
        public uint Sender()
        {
            return iSender;
        }
        public uint Receiver()
        {
            return iReceiver;
        }
        public uint Results()
        {
            return iResults;
        }
        protected override void CompleteRequest(IntPtr aAsyncHandle)
        {
            iService.EndPorts(aAsyncHandle, out iSender, out iReceiver, out iResults);
        }
    };

    /// <summary>
    /// Proxy for the av.openhome.org:NetworkMonitor:1 UPnP service
    /// </summary>
    public class CpProxyAvOpenhomeOrgNetworkMonitor1 : CpProxy, IDisposable, ICpProxyAvOpenhomeOrgNetworkMonitor1
    {
        private OpenHome.Net.Core.Action iActionName;
        private OpenHome.Net.Core.Action iActionPorts;
        private PropertyString iName;
        private PropertyUint iSender;
        private PropertyUint iReceiver;
        private PropertyUint iResults;
        private System.Action iNameChanged;
        private System.Action iSenderChanged;
        private System.Action iReceiverChanged;
        private System.Action iResultsChanged;
        private Mutex iPropertyLock;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <remarks>Use CpProxy::[Un]Subscribe() to enable/disable querying of state variable and reporting of their changes.</remarks>
        /// <param name="aDevice">The device to use</param>
        public CpProxyAvOpenhomeOrgNetworkMonitor1(CpDevice aDevice)
            : base("av-openhome-org", "NetworkMonitor", 1, aDevice)
        {
            OpenHome.Net.Core.Parameter param;
            List<String> allowedValues = new List<String>();

            iActionName = new OpenHome.Net.Core.Action("Name");
            param = new ParameterString("Name", allowedValues);
            iActionName.AddOutputParameter(param);

            iActionPorts = new OpenHome.Net.Core.Action("Ports");
            param = new ParameterUint("Sender");
            iActionPorts.AddOutputParameter(param);
            param = new ParameterUint("Receiver");
            iActionPorts.AddOutputParameter(param);
            param = new ParameterUint("Results");
            iActionPorts.AddOutputParameter(param);

            iName = new PropertyString("Name", NamePropertyChanged);
            AddProperty(iName);
            iSender = new PropertyUint("Sender", SenderPropertyChanged);
            AddProperty(iSender);
            iReceiver = new PropertyUint("Receiver", ReceiverPropertyChanged);
            AddProperty(iReceiver);
            iResults = new PropertyUint("Results", ResultsPropertyChanged);
            AddProperty(iResults);
            
            iPropertyLock = new Mutex();
        }

        /// <summary>
        /// Invoke the action synchronously
        /// </summary>
        /// <remarks>Blocks until the action has been processed
        /// on the device and sets any output arguments</remarks>
        /// <param name="aName"></param>
        public void SyncName(out String aName)
        {
            SyncNameAvOpenhomeOrgNetworkMonitor1 sync = new SyncNameAvOpenhomeOrgNetworkMonitor1(this);
            BeginName(sync.AsyncComplete());
            sync.Wait();
            sync.ReportError();
            aName = sync.Name();
        }

        /// <summary>
        /// Invoke the action asynchronously
        /// </summary>
        /// <remarks>Returns immediately and will run the client-specified callback when the action
        /// later completes.  Any output arguments can then be retrieved by calling
        /// EndName().</remarks>
        /// <param name="aCallback">Delegate to run when the action completes.
        /// This is guaranteed to be run but may indicate an error</param>
        public void BeginName(CallbackAsyncComplete aCallback)
        {
            Invocation invocation = iService.Invocation(iActionName, aCallback);
            int outIndex = 0;
            invocation.AddOutput(new ArgumentString((ParameterString)iActionName.OutputParameter(outIndex++)));
            iService.InvokeAction(invocation);
        }

        /// <summary>
        /// Retrieve the output arguments from an asynchronously invoked action.
        /// </summary>
        /// <remarks>This may only be called from the callback set in the above Begin function.</remarks>
        /// <param name="aAsyncHandle">Argument passed to the delegate set in the above Begin function</param>
        /// <param name="aName"></param>
        public void EndName(IntPtr aAsyncHandle, out String aName)
        {
            uint code;
            string desc;
            if (Invocation.Error(aAsyncHandle, out code, out desc))
            {
                throw new ProxyError(code, desc);
            }
            uint index = 0;
            aName = Invocation.OutputString(aAsyncHandle, index++);
        }

        /// <summary>
        /// Invoke the action synchronously
        /// </summary>
        /// <remarks>Blocks until the action has been processed
        /// on the device and sets any output arguments</remarks>
        /// <param name="aSender"></param>
        /// <param name="aReceiver"></param>
        /// <param name="aResults"></param>
        public void SyncPorts(out uint aSender, out uint aReceiver, out uint aResults)
        {
            SyncPortsAvOpenhomeOrgNetworkMonitor1 sync = new SyncPortsAvOpenhomeOrgNetworkMonitor1(this);
            BeginPorts(sync.AsyncComplete());
            sync.Wait();
            sync.ReportError();
            aSender = sync.Sender();
            aReceiver = sync.Receiver();
            aResults = sync.Results();
        }

        /// <summary>
        /// Invoke the action asynchronously
        /// </summary>
        /// <remarks>Returns immediately and will run the client-specified callback when the action
        /// later completes.  Any output arguments can then be retrieved by calling
        /// EndPorts().</remarks>
        /// <param name="aCallback">Delegate to run when the action completes.
        /// This is guaranteed to be run but may indicate an error</param>
        public void BeginPorts(CallbackAsyncComplete aCallback)
        {
            Invocation invocation = iService.Invocation(iActionPorts, aCallback);
            int outIndex = 0;
            invocation.AddOutput(new ArgumentUint((ParameterUint)iActionPorts.OutputParameter(outIndex++)));
            invocation.AddOutput(new ArgumentUint((ParameterUint)iActionPorts.OutputParameter(outIndex++)));
            invocation.AddOutput(new ArgumentUint((ParameterUint)iActionPorts.OutputParameter(outIndex++)));
            iService.InvokeAction(invocation);
        }

        /// <summary>
        /// Retrieve the output arguments from an asynchronously invoked action.
        /// </summary>
        /// <remarks>This may only be called from the callback set in the above Begin function.</remarks>
        /// <param name="aAsyncHandle">Argument passed to the delegate set in the above Begin function</param>
        /// <param name="aSender"></param>
        /// <param name="aReceiver"></param>
        /// <param name="aResults"></param>
        public void EndPorts(IntPtr aAsyncHandle, out uint aSender, out uint aReceiver, out uint aResults)
        {
            uint code;
            string desc;
            if (Invocation.Error(aAsyncHandle, out code, out desc))
            {
                throw new ProxyError(code, desc);
            }
            uint index = 0;
            aSender = Invocation.OutputUint(aAsyncHandle, index++);
            aReceiver = Invocation.OutputUint(aAsyncHandle, index++);
            aResults = Invocation.OutputUint(aAsyncHandle, index++);
        }

        /// <summary>
        /// Set a delegate to be run when the Name state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgNetworkMonitor1 instance will not overlap.</remarks>
        /// <param name="aNameChanged">The delegate to run when the state variable changes</param>
        public void SetPropertyNameChanged(System.Action aNameChanged)
        {
            lock (iPropertyLock)
            {
                iNameChanged = aNameChanged;
            }
        }

        private void NamePropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iNameChanged);
            }
        }

        /// <summary>
        /// Set a delegate to be run when the Sender state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgNetworkMonitor1 instance will not overlap.</remarks>
        /// <param name="aSenderChanged">The delegate to run when the state variable changes</param>
        public void SetPropertySenderChanged(System.Action aSenderChanged)
        {
            lock (iPropertyLock)
            {
                iSenderChanged = aSenderChanged;
            }
        }

        private void SenderPropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iSenderChanged);
            }
        }

        /// <summary>
        /// Set a delegate to be run when the Receiver state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgNetworkMonitor1 instance will not overlap.</remarks>
        /// <param name="aReceiverChanged">The delegate to run when the state variable changes</param>
        public void SetPropertyReceiverChanged(System.Action aReceiverChanged)
        {
            lock (iPropertyLock)
            {
                iReceiverChanged = aReceiverChanged;
            }
        }

        private void ReceiverPropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iReceiverChanged);
            }
        }

        /// <summary>
        /// Set a delegate to be run when the Results state variable changes.
        /// </summary>
        /// <remarks>Callbacks may be run in different threads but callbacks for a
        /// CpProxyAvOpenhomeOrgNetworkMonitor1 instance will not overlap.</remarks>
        /// <param name="aResultsChanged">The delegate to run when the state variable changes</param>
        public void SetPropertyResultsChanged(System.Action aResultsChanged)
        {
            lock (iPropertyLock)
            {
                iResultsChanged = aResultsChanged;
            }
        }

        private void ResultsPropertyChanged()
        {
            lock (iPropertyLock)
            {
                ReportEvent(iResultsChanged);
            }
        }

        /// <summary>
        /// Query the value of the Name property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the Name property</returns>
        public String PropertyName()
        {
            PropertyReadLock();
            String val;
            try
            {
                val = iName.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Query the value of the Sender property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the Sender property</returns>
        public uint PropertySender()
        {
            PropertyReadLock();
            uint val;
            try
            {
                val = iSender.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Query the value of the Receiver property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the Receiver property</returns>
        public uint PropertyReceiver()
        {
            PropertyReadLock();
            uint val;
            try
            {
                val = iReceiver.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Query the value of the Results property.
        /// </summary>
        /// <remarks>This function is threadsafe and can only be called if Subscribe() has been
        /// called and a first eventing callback received more recently than any call
        /// to Unsubscribe().</remarks>
        /// <returns>Value of the Results property</returns>
        public uint PropertyResults()
        {
            PropertyReadLock();
            uint val;
            try
            {
                val = iResults.Value();
            }
            finally
            {
                PropertyReadUnlock();
            }
            return val;
        }

        /// <summary>
        /// Must be called for each class instance.  Must be called before Core.Library.Close().
        /// </summary>
        public void Dispose()
        {
            lock (this)
            {
                if (iHandle == IntPtr.Zero)
                    return;
                DisposeProxy();
                iHandle = IntPtr.Zero;
            }
            iActionName.Dispose();
            iActionPorts.Dispose();
            iName.Dispose();
            iSender.Dispose();
            iReceiver.Dispose();
            iResults.Dispose();
        }
    }
}

